File: optimizer/objectives/runtime_objective.py
"""Extract one auditable latency objective from an evidence report."""

from __future__ import annotations

from typing import Any


def runtime_objective(
    report: dict[str, Any], *, metric: str, quantile: str
) -> dict[str, Any]:
    metrics = report.get("metrics_ns")
    values = metrics.get(metric) if isinstance(metrics, dict) else None
    if not isinstance(values, dict) or quantile not in values:
        raise ValueError(f"missing {metric} {quantile}")
    value = values[quantile]
    if isinstance(value, bool) or not isinstance(value, (int, float)) or value < 0:
        raise ValueError(f"invalid {metric} {quantile}")
    rate = report.get("complete_trace_rate")
    if rate is None:
        complete = report.get("complete_trace_count")
        observed = report.get("observed_trace_count")
        if (
            isinstance(complete, int)
            and not isinstance(complete, bool)
            and isinstance(observed, int)
            and not isinstance(observed, bool)
            and observed > 0
            and 0 <= complete <= observed
        ):
            rate = complete / observed
    if (
        isinstance(rate, bool)
        or not isinstance(rate, (int, float))
        or not 0 <= rate <= 1
    ):
        raise ValueError("invalid complete_trace_rate")
    formal = (
        report.get("development_only") is False
        and report.get("formal_inference_allowed") is True
    )
    return {
        "schema_version": "runtime-objective/v1",
        "source_schema_version": report.get("schema_version", ""),
        "metric": metric,
        "quantile": quantile,
        "objective_value_ns": float(value),
        "complete_trace_rate": float(rate),
        "formal_optimization_allowed": formal,
    }

File: cpp_core/optimizer/src/runtime_objective.cpp
#include "robotraceopt/optimizer/runtime_objective.hpp"

#include <cmath>
#include <stdexcept>

namespace robotraceopt::optimizer {

RuntimeObjective runtime_objective(
    const RuntimeReport& report,
    const std::string& metric,
    const std::string& quantile) {
    const auto metric_it = report.metrics_ns.find(metric);
    if (metric_it == report.metrics_ns.end()) {
        throw std::invalid_argument("missing " + metric + " " + quantile);
    }
    const auto quantile_it = metric_it->second.find(quantile);
    if (quantile_it == metric_it->second.end()) {
        throw std::invalid_argument("missing " + metric + " " + quantile);
    }
    const double value = quantile_it->second;
    if (!std::isfinite(value) || value < 0.0) {
        throw std::invalid_argument("invalid " + metric + " " + quantile);
    }

    double rate = 0.0;
    if (report.complete_trace_rate.has_value()) {
        rate = *report.complete_trace_rate;
    } else if (
        report.complete_trace_count.has_value() &&
        report.observed_trace_count.has_value() &&
        *report.observed_trace_count > 0 &&
        *report.complete_trace_count <= *report.observed_trace_count) {
        rate = static_cast<double>(*report.complete_trace_count) /
            static_cast<double>(*report.observed_trace_count);
    } else {
        throw std::invalid_argument("invalid complete_trace_rate");
    }
    if (!std::isfinite(rate) || rate < 0.0 || rate > 1.0) {
        throw std::invalid_argument("invalid complete_trace_rate");
    }

    const bool formal =
        report.development_only.has_value() &&
        !*report.development_only &&
        report.formal_inference_allowed.has_value() &&
        *report.formal_inference_allowed;
    return RuntimeObjective{
        report.schema_version,
        metric,
        quantile,
        value,
        rate,
        formal,
    };
}

}  // namespace robotraceopt::optimizer
